source/hwIo/ioThread.py

```diff
@@ -10,176 +10,245 @@
+	#: This allows us to have a single APC function in the class rather than on
+	#: each instance, which prevents reference cycles.
+	_apcStore: ApcStoreT = {}
+	#: Store of Python functions to be called as Overlapped Completion Routine.
+	#: This allows us to have a single completion routine in the class rather than on
+	#: each instance, which prevents reference cycles.
+	#: Note that we use the address of the OVERLAPPED structure as key in this store,
+	#: eventhough the structure is also stored in the value.
+	#: The OVERLAPPED structure can't be used as key because ctypes does not have OOR (original object return),
+	#: it constructs a new, equivalent object each time you retrieve the contents of a LPOVERLAPPED.
+	_completionRoutineStore: CompletionRoutineStoreTypeT = {}
 
 	def __init__(self):
 		super().__init__(
 			name=f"{self.__class__.__module__}.{self.__class__.__qualname__}",
 			daemon=True
 		)
-		self._apcReferences = dict()
+
+	@winKernel.PAPCFUNC
+	def _internalApc(param: ApcIdT):
+		threadinst = threading.current_thread()
+		if not isinstance(threadinst, IoThread):
+			log.error("Internal APC called from unknown thread")
+			return
+
+		(reference, actualParam) = IoThread._apcStore.pop(param, (None, 0))
+		if reference is None:
+			log.error(f"Internal APC called with param {param}, but no such apcId in store")
+			return
+		if isinstance(reference, (BoundMethodWeakref, AnnotatableWeakref)):
+			function = reference()
+			if not function:
+				log.debugWarning(
+					f"Not executing queued APC {param}:{reference.funcName} with param {actualParam} because reference died"
+				)
+				return
+		else:
+			function = reference
+
+		try:
+			function(actualParam)
+		except Exception:
+			log.error(f"Error in APC function {function!r} with apcId {param} queued to IoThread", exc_info=True)
+
+	@LPOVERLAPPED_COMPLETION_ROUTINE
+	def _internalCompletionRoutine(
+			error: int,
+			numberOfBytes: int,
+			overlapped: LPOVERLAPPED
+	):
+		threadinst = threading.current_thread()
+		if not isinstance(threadinst, IoThread):
+			log.error("Internal APC called from unknown thread")
+			return
+
+		ptr = ctypes.cast(overlapped, ctypes.c_void_p).value
+		(reference, cachedOverlapped) = IoThread._completionRoutineStore.pop(ptr, (None, None))
+		if reference is None:
+			log.error(f"Internal completion routine called with pointer 0x{ptr:x}, but no such address in store")
+			return
+
+		function = reference()
+		if not function:
+			log.debugWarning(
+				f"Not executing queued completion routine 0x{ptr:x}:{reference.funcName} because reference died"
+			)
+			return
+
+		try:
+			function(error, numberOfBytes, overlapped)
+		except Exception:
+			log.error(f"Error in overlapped completion routine {function!r}", exc_info=True)
 
 	def start(self):
 		super().start()
 		self.handle = ctypes.windll.kernel32.OpenThread(winKernel.THREAD_SET_CONTEXT, False, self.ident)
 
-	@contextmanager
-	def autoDeleteApcReference(self, apcUuid: uuid.UUID):
```

Comment:
removing this would be an API breaking change, correct?

Reply:
Mm, strictly spoken may be. I should have made this an underscore prefixed method, as it is really meant to be internal to the ioThread. I really doubt that there will be an add-on that started using this post 2023.1 as this was added in that release.

Reply:
By our own policies and add-on author expectations, we need to prevent API breaking changes

Reply:
I'll see what I can do to avoid removing this. The problem is that in this pr, we no longer store apcs in a cache.
Could you agree with reintroducing this function but making it a logging no-op marked as API breaking?

Reply:
Yes - logging at error/exception level as a no-op is an improvement than a AttributeError being raised.